model/private_subnet.rb

```diff
@@ -90,4 +100,37 @@ def random_private_ipv6
 
     addr
   end
+
+  def connect_subnet(subnet)
+    small_id_ps, large_id_ps = [self, subnet].sort_by(&:id)
+    ConnectedSubnet.create_with_id(subnet_id_1: small_id_ps.id, subnet_id_2: large_id_ps.id)
+    nics.each do |nic|
+      create_tunnels(subnet.nics, nic)
+    end
+    subnet.incr_refresh_keys
+  end
+
+  def disconnect_subnet(subnet)
+    small_id_ps, large_id_ps = [self, subnet].sort_by(&:id)
+    nics.each do |nic|
+      (nic.src_ipsec_tunnels + nic.dst_ipsec_tunnels).each do |tunnel|
+        tunnel.destroy if tunnel.src_nic.private_subnet_id == subnet.id || tunnel.dst_nic.private_subnet_id == subnet.id
+      end
+    end
+    ConnectedSubnet.where(subnet_id_1: small_id_ps.id, subnet_id_2: large_id_ps.id).all.map(&:destroy)
+    subnet.incr_refresh_keys
+    incr_refresh_keys
+  end
+
+  def create_tunnels(nics, src_nic)
+    nics.each do |dst_nic|
+      next if src_nic == dst_nic
+      IpsecTunnel.create_with_id(src_nic_id: src_nic.id, dst_nic_id: dst_nic.id) unless IpsecTunnel[src_nic_id: src_nic.id, dst_nic_id: dst_nic.id]
+      IpsecTunnel.create_with_id(src_nic_id: dst_nic.id, dst_nic_id: src_nic.id) unless IpsecTunnel[src_nic_id: dst_nic.id, dst_nic_id: src_nic.id]
```

Comment:
This is a bit smelly, considering that create_with_id is still generally entitled to fail with a unique constraint violation anyway. One option might be "insert...on conflict..."

Reply:
This was used as is before, too. I only changed its place, I'll create an item to look into it later.